feat: add `blas/ext/base/dcartesian-square` (#10797)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/ext/base/dcartesian-square/README.md

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
var Float64Array = require( '@stdlib/array/float64' );
var dcartesianSquare = require( '@stdlib/blas/ext/base/dcartesian-square' );

var N = 4;
var x = discreteUniform( N, 1, 10, {
    'dtype': 'float64'
});
console.log( x );

var out = new Float64Array( N * N * 2 );
dcartesianSquare( N, x, 1, out, 2, 1 );
console.log( out );
```

</section>

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/blas/ext/base/dcartesiansquare.h"
```

#### stdlib_strided_dcartesian_square( N, \*X, strideX, \*Out, strideOut1, strideOut2 )

Computes the Cartesian square for a double-precision floating-point strided array.

```c

> double x[] = { 1.0, 2.0, 3.0, 4.0 };

Author: ```suggestion
const double x[] = { 1.0, 2.0, 3.0, 4.0 };
```

double out[ 32 ];

stdlib_strided_dcartesian_square( 4, x, 1, out, 2, 1 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **X**: `[in] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **Out**: `[out] double*` output array.
-   **strideOut1**: `[in] CBLAS_INT` stride length between consecutive output pairs.
-   **strideOut2**: `[in] CBLAS_INT` stride length between elements of each output pair.

```c
void stdlib_strided_dcartesian_square( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, double *Out, const CBLAS_INT strideOut1, const CBLAS_INT strideOut2 );
```

<!-- lint disable maximum-heading-length -->

#### stdlib_strided_dcartesian_square_ndarray( N, \*X, strideX, offsetX, \*Out, strideOut1, strideOut2, offsetOut )

Computes the Cartesian square for a double-precision floating-point strided array using alternative indexing semantics.

```c

> double x[] = { 1.0, 2.0, 3.0, 4.0 };

Author: ```suggestion
const double x[] = { 1.0, 2.0, 3.0, 4.0 };
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **X**: `[in] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
-   **Out**: `[out] double*` output array.
-   **strideOut1**: `[in] CBLAS_INT` stride length between consecutive output pairs.
-   **strideOut2**: `[in] CBLAS_INT` stride length between elements of each output pair.
-   **offsetOut**: `[in] CBLAS_INT` starting index for `Out`.

```c
void stdlib_strided_dcartesian_square_ndarray( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, double *Out, const CBLAS_INT strideOut1, const CBLAS_INT strideOut2, const CBLAS_INT offsetOut );
```

</section>

<!-- /.usage -->

<!-- C API usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

</section>

<!-- /.notes -->

<!-- C API usage examples. -->

<section class="examples">

### Examples

```c
#include "stdlib/blas/ext/base/dcartesiansquare.h"
#include <stdio.h>

int main( void ) {
    // Create a strided input array:

>     double x[] = { 1.0, 2.0, 3.0, 4.0 };

Author: ```suggestion
    const double x[] = { 1.0, 2.0, 3.0, 4.0 };
```

-   **order**: storage layout. Must be either `'row-major'` or `'column-major'`.
-   **N**: number of indexed elements.
-   **x**: input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: stride length for `x`.
-   **out**: output [`Float64Array`][@stdlib/array/float64].
-   **LDO**: stride length for the leading dimension of `out`.

The `N` and stride parameters determine which elements in the strided arrays are accessed at runtime. For example, to compute the Cartesian square of every other element:

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 1.0, 0.0, 2.0, 0.0 ] );
var out = new Float64Array( 8 );

dcartesianSquare( 'row-major', 2, x, 2, out, 2 );
// out => <Float64Array>[ 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0 ]
```

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

// Initial array...
var x0 = new Float64Array( [ 0.0, 1.0, 2.0, 3.0 ] );

// Create an offset view...
var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

> 

Reviewer: ```suggestion
// Initial array:
var x0 = new Float64Array( [ 0.0, 1.0, 2.0, 3.0 ] );

// Create an offset view:
var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

// Output array:
```

var out = new Float64Array( 8 );

dcartesianSquare( 'row-major', 2, x1, 1, out, 2 );
// out => <Float64Array>[ 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0 ]
```

<!-- lint disable maximum-heading-length -->

#### dcartesianSquare.ndarray( N, x, strideX, offsetX, out, strideOut1, strideOut2, offsetOut )

<!-- lint enable maximum-heading-length -->

Computes the Cartesian square for a double-precision floating-point strided array using alternative indexing semantics.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 1.0, 2.0 ] );
var out = new Float64Array( 8 );

dcartesianSquare.ndarray( x.length, x, 1, 0, out, 2, 1, 0 );
// out => <Float64Array>[ 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0 ]
```

> The function has the following additional parameters:

Reviewer: We should document all the parameters here. No longer does the method have just additional parameters, but it has differently named parameters.

-   **offsetX**: starting index for `x`.
-   **strideOut1**: stride length of the first dimension of `out`.
-   **strideOut2**: stride length of the second dimension of `out`.
-   **offsetOut**: starting index for `out`.

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to access only the last two elements:

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 0.0, 0.0, 1.0, 2.0 ] );
var out = new Float64Array( 8 );

dcartesianSquare.ndarray( 2, x, 1, 2, out, 2, 1, 0 );
// out => <Float64Array>[ 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0 ]
```

</section>

<!-- /.usage -->

<section class="notes">

> ## Notes

Reviewer: You need to add a note indicating how the pairs are stored. Namely, they are stored as rows in the output matrix. We should also add a similar comment to the `repl.txt` file, the TypeScript declaration file, and the `lib/ndarray.js` and `main.c` files. That is pertinent information needed by end users to interpret the results.

> -   For an input array `x` of length `N`, the output array `out` must contain at least `N^2 * LDO` elements.

Reviewer: ```suggestion
-   For an input array of length `N`, the output array must contain at least `N * N * 2` indexed elements.
```

```c
void stdlib_strided_dcartesian_square_ndarray( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, double *Out, const CBLAS_INT strideOut1, const CBLAS_INT strideOut2, const CBLAS_INT offsetOut );
```

</section>

<!-- /.usage -->

<!-- C API usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

</section>

<!-- /.notes -->

<!-- C API usage examples. -->

<section class="examples">

### Examples

```c
#include "stdlib/blas/ext/base/dcartesiansquare.h"
#include "stdlib/blas/base/shared.h"
#include <stdio.h>

int main( void ) {
    // Create a strided input array:
    const double x[] = { 1.0, 2.0 };

    // Specify the number of indexed elements:
    const int N = 2;

    // Create an output array (N*N pairs, each pair has 2 elements):
    double out[ 8 ];

    // Specify strides:
    const int strideX = 1;
    const int LDO = 2;

    // Compute the Cartesian square:
    stdlib_strided_dcartesian_square( CblasRowMajor, N, x, strideX, out, LDO );

    // Print the result:
    for ( int i = 0; i < N*N; i++ ) {

>         printf( "out[ %i ] = ( %lf, %lf )\n", i, out[ i*2 ], out[ i*2+1 ] );

Reviewer: ```suggestion
        printf( "out[ %i ] = ( %lf, %lf )\n", i, out[ i*2 ], out[ (i*2)+1 ] );
```

WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dcartesianSquare

> Compute the Cartesian square for a double-precision floating-point strided array.

<section class="usage">

## Usage

```javascript
var dcartesianSquare = require( '@stdlib/blas/ext/base/dcartesian-square' );
```

#### dcartesianSquare( order, N, x, strideX, out, LDO )

Computes the Cartesian square for a double-precision floating-point strided array.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 1.0, 2.0 ] );
var out = new Float64Array( 8 );

dcartesianSquare( 'row-major', x.length, x, 1, out, 2 );
// out => <Float64Array>[ 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0 ]
```

The function has the following parameters:

-   **order**: storage layout. Must be either `'row-major'` or `'column-major'`.
-   **N**: number of indexed elements.
-   **x**: input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: stride length for `x`.
-   **out**: output [`Float64Array`][@stdlib/array/float64].

> -   **LDO**: stride length for the leading dimension of `out`.

Reviewer: ```suggestion
-   **LDO**: stride length between successive contiguous vectors of the matrix `out` (a.k.a., leading dimension of `out`).
```

-   **strideX**: stride length for `x`.
-   **offsetX**: starting index for `x`.
-   **out**: output [`Float64Array`][@stdlib/array/float64].
-   **strideOut1**: stride length of the first dimension of `out`.
-   **strideOut2**: stride length of the second dimension of `out`.
-   **offsetOut**: starting index for `out`.

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to access only the last two elements:

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 0.0, 0.0, 1.0, 2.0 ] );
var out = new Float64Array( 8 );

dcartesianSquare.ndarray( 2, x, 1, 2, out, 2, 1, 0 );
// out => <Float64Array>[ 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0 ]
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   Pairs are stored as rows in the output matrix, where the first column contains the first element of each pair and the second column contains the second element.
-   For an input array of length `N`, the output array must contain at least `N * N * 2` indexed elements.

> -   For row-major order, the `LDO` parameter must be greater than or equal to `max(1,2)`. For column-major order, the `LDO` parameter must be greater than or equal to `max(1,N*N)`.

Reviewer: ```suggestion
-   For row-major order, the `LDO` parameter must be greater than or equal to `2`. For column-major order, the `LDO` parameter must be greater than or equal to `max(1,N*N)`.
```

var out = new Float64Array( N * N * 2 );
dcartesianSquare( 'row-major', N, x, 1, out, 2 );
console.log( out );
```

</section>

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/blas/ext/base/dcartesiansquare.h"
```

#### stdlib_strided_dcartesian_square( order, N, \*X, strideX, \*Out, LDO )

Computes the Cartesian square for a double-precision floating-point strided array.

```c
#include "stdlib/blas/base/shared.h"

const double x[] = { 1.0, 2.0 };
double out[ 8 ];

stdlib_strided_dcartesian_square( CblasRowMajor, 2, x, 1, out, 2 );
```

The function accepts the following arguments:

-   **order**: `[in] CBLAS_LAYOUT` storage layout.
-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **X**: `[in] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **Out**: `[out] double*` output array.

> -   **LDO**: `[in] CBLAS_INT` stride of the leading dimension of `Out`. For row-major order, must be greater than or equal to `max(1,2)`. For column-major order, must be greater than or equal to `max(1,N*N)`.

Reviewer: ```suggestion
-   **LDO**: `[in] CBLAS_INT` stride length between successive contiguous vectors of the matrix `Out` (a.k.a., leading dimension of `Out`). For row-major order, must be greater than or equal to `max(1,2)`. For column-major order, must be greater than or equal to `max(1,N*N)`.
```

> -   **LDO**: `[in] CBLAS_INT` stride length between successive contiguous vectors of the matrix `Out` (a.k.a., leading dimension of `Out`). For row-major order, must be greater than or equal to `max(1,2)`. For column-major order, must be greater than or equal to `max(1,N*N)`.

Reviewer: ```suggestion
-   **LDO**: `[in] CBLAS_INT` stride length between successive contiguous vectors of the matrix `Out` (a.k.a., leading dimension of `Out`). For row-major order, must be greater than or equal to `2`. For column-major order, must be greater than or equal to `max(1,N*N)`.
```
